ADR020 on improving arena edges (#1770)
informalsystems/apalache: docs/src/adr/020adr-arenas.md

---
authors: Igor Konnov
last revised: 2022-05-16
---

# ADR-020: Improving membership in arenas

**Table of Contents**

- [Summary (Overview)](#summary)
- [Context (Problem)](#context)
- [Options (Alternatives)](#options)
- [Solution (Decision)](#solution)
- [Consequences (Retrospection)](#consequences)

## Summary

> We discuss an extension of membership edges in arenas. The main application of

Reviewer: I think the way this is framed as an extension is misleading.
Isn't it more about lifting points-to information from the SMT encoding / SMT membership constants into the rewriter?

Author: Yes, arenas are also some form of static points-to information, though it is not as precise, as it could be

Reviewer: I see. I didn't really get this from reading the summary/context. Maybe we can add something along the lines of

1. Static points-to information currently present in the SMT encoding is not as precise as it could be.
2. We propose to improve this by providing more accurate information on the level of arenas.
3. We achieve this via an extension of membership edges proposed here.

to the context section?

and sequences) are translated into *cells*. Cells are SMT constants, which can
be connected to other cells. Cells are connected with one another by the means
of edges. Currently, we have three kinds of edges:

 - **has**. A membership edge `(c_p, c_e)` represents that a parent cell `c_p`
   potentially contains an element `c_e` (e.g., if `c_p` represents a set).
   These edges encode many-to-many relations.

 - **dom**. A domain edge `(c_f, c_d)` represents that a function cell `c_f`
   has the domain represented with a cell `c_d`.
   These edges encode many-to-one relations.

 - **cdm**. A co-domain edge `(c_F, c_c)` represents that a function set cell
   `c_F` has the co-domain represented with a cell `c_c`.
   These edges encode many-to-one relations.

   *For historic reasons, functions are also encoded with the edges called
   **dom** and **cdm**, though the **cdm** edge points to the function
   relation, not its co-domain. This should be fixed!*

Having the edges of kinds **dom** and **cdm** happened to be two inflexible.
As we explained above, sometimes we would like to have another edge kind called

> **rel**. As these edges are many-to-one, we can map them from their kinds `kind

Author: added an explanation

-> (c_p, c_e)`. This requires simple refactoring, so we are not going to
discuss the **dom** and **cdm** any further.

There is a need for refactoring and extension of the **has**-edges. We
summarize the issues with the current implementation of this kind of edges:

 - Originally, every edge `(c_S, c_e)` of the kind **has** was encoded as a
   Boolean constant `in_${c_e.id}_${c_S.id}` in SMT. Hence, every time the edge
   is copied to another set cell `c_T`, we have to introduce another Boolean
   constant `in_${c_e.id}_${c_T.id}` in SMT. As a result, two set cells do not

>    share edges. This causes unnecessary duplication of SMT constants and

Author: added an explanation

   constraints.

 - Later, we stopped introducing Boolean constants in SMT for the **has**-edges
   when translating records and tuples. However, we do not record the fact that
   these edges are presented only in the arena, not in SMT. Hence, we have to
   be careful and do not use the SMT membership when working with these edges.

 - As every **has**-edge directly refers to its parent in the edge name (that
   is, `in_${c_e.id}_${c_S.id}`), we cannot share edges when encoding `SUBSET
   S` and `[S -> T]`. As a result, we have to introduce a massive number of
   Boolean constants and constraints, which are not necessary.

 - We keep adding edges and SMT constants to the solver context, even when we
   know exactly that an element belongs to a set, e.g., as in `{ 1, 2, 3 }`.

## 2. Options

 - Keep things as they are.

>  - Extend the edges.

Reviewer: As I said above: I think framing as an extension is misleading.

TLA+ IR:

```scala
/**
 * An element pointer whose value is encoded via an SMT expression.
 * @param elem
 *   the element this pointer is pointing to.
 * @param smtEx
 *   the corresponding SMT expression.
 */
case class SmtExprElemPtr(elem: ArenaCell, smtEx: TlaEx) extends ElemPtr {
  override def toSmt: TlaEx = smtEx
}
```

Instances of `SmtExprElemPtr` may be used to encode set membership via SMT
expressions. For instance:

 - Evaluating an array expression, e.g., via `apalacheStoreInFun`.

 - Combining several pointers. For instance, when computing `{ x \in S: P }`,
   we would combine set membership in `S` and the value of `P` for every `x`.

### 3.2. Optimization 1: constant propagation via membership pointers

One immediate application of using the new representation is completely
SMT-free construction of some of the TLA+ expressions. Consider the following
expression:

```tla
{ a, b, c } \union { d, e }
```

Let's denote the arguments of the set union to be `S` and `T`. In the current
arena representation, the rewriting rule `SetCtorRule` creates the following
SMT constants (assuming that `a, ..., e` were translated to

> arena cells):

Author: Yep, moved this example to the context

   constant of the sort of `S` and `T`.

 - One cell `c_u` to represent the set `S \union T`.

 - Five Boolean constants of the Boolean sort that express set membership of
   `a, b, c, d, e` in `S \union T`.

It is obvious that 10 Boolean constants introduced for set membership are
completely unnecessary, as we know for sure that the respective elements belong
to the three sets. Moreover, when constructing `S \union T`, the rule `SetCupRule`
creates five SMT constraints:

```smt
;; a, b, c belong to the union, when they belong to S
(= in_a_u in_a_l)
(= in_b_u in_b_l)
(= in_c_u in_c_l)
;; d and e belong to the union, when they belong to T
(= in_d_u in_d_r)
(= in_e_u in_e_r)
```

With the new representation, the set constructor would simply create five
instances of `FixedElemPtr` that carry the value `true`, that is, the elements
unconditionally belong to `S` and `T`. Further, the rule `SetCupRule` would
simply copy the five pointers, without propagating anything to SMT.

As a result, we obtain constant propagation of set membership, while
keeping the general spirit of the arena-based encoding.

### 3.3. Optimization 2: sharing membership in a powerset

> Consider the TLA+ operator that construct the powerset of `S`, that is, the set

Reviewer 2: ```suggestion
Consider the TLA+ operator that constructs the powerset of `S`, that is, the set
```

of all subsets of `S`:

```tla
SUBSET S
```

Let `c_S` be the cell that represents the set `S` and `c_1, ..., c_n` be the
cells that represent the potential elements of `S`. Note that in general,

> membership of all these cells may be unknown in static. For example, consider

Reviewer 2: ```suggestion
membership of all these cells may be statically unknown. For example, consider
```

subsets of `S`. The tricky part here is that some of the elements of `S` may be
outside of `S`. To deal with that, `PowSetCtor` constructs cells for each
potential combinations of `c_1, ..., c_n` and adds membership tests for each of
them. For instance, consider the subset `T` that is constructed by selecting the
indices `1, 3, 5` of `1..n`. The constructor will introduce `3`
constraints:

```smt
(= in_c_1_T in_c_1_S)
(= in_c_3_T in_c_3_S)
(= in_c_5_T in_c_5_S)
```

Hence, the current encoding introduces `2^n` SMT constants for the subsets and
`n * 2^(n - 1)` membership constraints in SMT (thanks to Jure @Kukovec for
telling me the precise formula).

With the new representation, we would simply copy the respective membership
pointers of the set `S`. This would require us to introduce zero constraints in
the SMT, though we would still introduce `2^n` SMT constants to represent the
subsets themselves. Note that we would still have to introduce `n * 2^(n - 1)`
pointers in the arena. But this would be done during the process of rewriting.

### 3.4. Feature: computing the set of functions via pointer sharing

Sometimes, it happens that the model checker has to expand a set of functions

> `[S -> T]`. Such a set contains `|T|^|S|` functions. Since the model checker is
> working with arenas, it can only construct an arena representation of `[S ->

Reviewer 2: ```suggestion
`[S -> T]`. Such a set contains `|T|^|S|` functions. Since the model checker works with arenas, it can only construct an arena representation of `[S ->
```

T]`. To this end, assume that the set `S` is encoded via cells `s_1, ..., s_m`,
and the set `T` is encoded via cells `t_1, ..., t_n`.

> Had we constructed `[S -> T]` in the current encoding, we would have to

Reviewer 2: ```suggestion
If we wanted to construct `[S -> T]` in the current encoding, we would have to
```

introduce a relation for each function in the set `[S -> T]`. That is, for
every sequence `i[1], ..., i[n]` over `1..n`, it would construct the relation
`R`:

    { <<s_1, t_i[1]>>, ..., <<s_m, t_i[m]>> }

Let's denote with `p_j` the pair `<<s_j, t_i[j]>>` for `1 <= j <= m`.

Moreover, we would add `m` membership constraints (per function!) in SMT:

    (= in_p_1_R (and in_s_1_S in_t_i[1]_T))
    ...
    (= in_p_m_R (and in_s_m_S in_t_i[m]_T))

As a result, this encoding would introduce `m * n^m` constants in SMT and the
same number of membership constraints. For instance, if we have `m = 10` and `n
= 5`, then we would introduce 90 million constants and constraints!

Using the approach outlined in this ADR, we can simply combine membership
pointers of `S` and `T` via `SmtExprElemPtr`. This would neither introduce SMT

> constants, nor additional constraints. Of course, when this set is used in

Reviewer 2: How are "additional" constraints different from "SMT" constraints?
